src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,51 +27,125 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) {
+    const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(base_segment);
```

Comment:
Can we somehow DebugAssert this? That it is a ValueSegment? (I don't know if it is possible)

Reply:
`std::dynamic_pointer_cast` throws if the cast can not be done (checks vtable)

Reply:
Okay, so it might make sense to add `DebugAssert(value_segment, "Invalid base segment passed to dictionary segment constructor")` here. The dynamic cast will not fail, but it will return an empty shared pointer if the value can not safely be casted.